[#65043] Bump ViewComponent to 4.0.2, Primer to 0.72.0

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/65043

# What are you trying to accomplish?

* Bumps ViewComponent to 4.0.2
* Also bumps Primer to 0.72.0
  * Renames `Primer::OpenProject::TreeView` to `Primer::Alpha::TreeView`

⚠️ This PR contains a fix to ensure Sprockets is initialized before ViewComponent, working around asset resolution failures in test environment (see ViewComponent/view_component#2433).

# What approach did you choose and why?

For reference: 

- [Diff](https://github.com/ViewComponent/view_component/compare/main...v4)
- [CHANGELOG](https://github.com/ViewComponent/view_component/blob/v4/docs/CHANGELOG.md#400alpha6)

# Merge checklist

- [X] Fix upstream issues:
  - [X] https://github.com/ViewComponent/view_component/issues/2433 (workaround found)
  - [x] https://github.com/ViewComponent/view_component/issues/2359
    - [x] PR created
    - [x] PR merged
    - [x] Version updated
  - [ ] https://github.com/ViewComponent/view_component/issues/2413 
- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
Bump ViewComponent to 4.0.2

Fixes obsolete ViewComponent config in initializers.

See [CHANGELOG.md](https://github.com/ViewComponent/view_component/blob/main/docs/CHANGELOG.md#400rc1).

myabc committed:
Fix super calls in component #initialize methods

`ViewComponent::Base` previously inherited from `ActionView::Base` but
defined a catch-all `#initialize` to allow passing arbitary arguments.

See: https://github.com/ViewComponent/view_component/pull/2293
See: https://github.com/ViewComponent/view_component/pull/2287

myabc committed:
Fix erroneous calls to #title

These previously succeeded as meta-tags defines `#title` in
`MetaTags::ViewHelper` and mixes it in to `ActionView::Base`.

myabc committed:
Fix missing #render_widget method

myabc committed:
Bump Primer to 0.72.0

myabc committed:
Ensure Sprockets initialized before ViewComponent

Works around an issue where asset resolution fails in test environment.

     Sprockets::Rails::Helper::AssetNotFound:
       The asset "loading_indicator.svg" is not present in the asset pipeline.

See: ViewComponent/view_component#2433

myabc committed:
Rename constant to Primer::Alpha::TreeView

Per BREAKING CHANGE in Primer View Components 0.72.0.

This component has now been accepted into the upstream Primer project.
See primer/view_components#3563.

myabc commented on lib/open_project/patches/primer_page_header_breadcrumb.rb:
```diff
@@ -40,6 +40,6 @@ def with_breadcrumbs(breadcrumbs, skip_home_on_mobile: false, **)
 end
 
-OpenProject::Patches.patch_gem_version "openproject-primer_view_components", "0.71.0" do
+OpenProject::Patches.patch_gem_version "openproject-primer_view_components", "0.72.0" do
   Primer::OpenProject::PageHeader.prepend OpenProject::Patches::PrimerPageHeaderBreadcrumb
 end
```
@HDinger is this patch still necessary?

HDinger replied:
Yes, it is (and will most likely be for a long time). This ensures, that the instance name is always prepended to the Breadcrumb.

HDinger commented on modules/reporting/app/components/cost_reports/index_page_header_component.rb:
```diff
@@ -31,6 +31,7 @@
 module CostReports
   class IndexPageHeaderComponent < ApplicationComponent
     include ApplicationHelper
+    include Widget::ReportingWidget::RenderWidgetInstanceMethods
```
What is that needed for?

myabc replied:
This mixes in the `#render_widget` method.

HDinger reviewed (commented):
Changes look good as far as I can tell. There are however still quite some tests failing

myabc committed:
Simplify Forms #html_content implementation

This change also resolves issues with block scope after upgrading to
ViewComponent 4.0:

     ActionView::Template::Error:
       undefined local variable or method 'model' for an instance of #<Class:0x00000003020e3cb0>
     # ./app/forms/scim_clients/form.rb:93:in 'block (2 levels) in <class:Form>'
     # ./lib/primer/open_project/forms/html_content.rb:15:in 'BasicObject#instance_exec'

Additionally, since we are only interested in the passed `html_block`,
there is no need define a separate `HtmlContentInput` class.

myabc committed:
Fix PageHeader title, breadcrumbs double escaping

myabc committed:
Fix missing helpers proxy in LifeCycle::Form

`helpers` is not available in `Primer::Forms::Base`/`ActsAsComponent`
descendants.

myabc committed:
Small BorderBoxTableComponent optimizations

myabc committed:
Fix "lazy" translation scoping in components

See: ViewComponent/view_component#2413

myabc committed:
Skeleton specs for BorderBoxTableComponent impls.

myabc committed:
Remove duplicate RSpec config for component specs

myabc committed:
Do not include SystemTestHelpers in component specs

`SystemSpecHelpers` should be used.

myabc committed:
Fix TimeTracking components parent class

These components do not use the functionality provided by
`OpPrimer::BorderBoxTableComponent`. Most components should inherit from
`ApplicationComponent`.

oliverguenther reviewed (approved):
The two hacks look acceptable to me 👍 Let's try and get it in to avoid more conflicts.
